app_dart/lib/src/request_handling/single_execution_request_handler.dart

```diff
@@ -0,0 +1,113 @@
+/// > rare circumstances, it is possible for multiple instances of the same job
+/// > to be requested, therefore, your request handler should be idempotent, and
+/// > your code should ensure that there are no harmful side-effects if this
+/// > occurs.
+///
+/// [1]: https://cloud.google.com/appengine/docs/flexible/scheduling-jobs-with-cron-yaml
+abstract class SingleExecutionRequestHandler extends ApiRequestHandler {
+  @visibleForTesting
+  static const subCacheName = 'SingleExecutionRequestHandler';
+
+  @visibleForTesting
+  static const xAppengineCron = 'X-Appengine-Cron';
+
+  const SingleExecutionRequestHandler({
+    required super.config,
+    required super.authenticationProvider,
+    required CacheService cache,
+    @visibleForTesting DateTime Function() now = DateTime.now,
+  }) : _cache = cache,
+       _now = now;
+
+  final CacheService _cache;
+  final DateTime Function() _now;
+
+  /// Whether to disallow access from non-`cron.yaml` jobs.
+  ///
+  /// See <https://cloud.google.com/appengine/docs/flexible/scheduling-jobs-with-cron-yaml#securing_urls_for_cron>.
+  bool get allowOnlyAppEngineCronAccess => true;
+
+  /// The maximum amount of time this task might concievably run.
+  ///
+  /// After [maxExecutionTime], the "lock" is removed and subsequent requests
+  /// still invoke [get], even if (somehow) an existing handler is still
+  /// processing a job.
+  ///
+  /// At most 60 minutes (cron tasks cannot run longer than 60 minutes), but
+  /// often shorter, such as the duration set in the root `cron.yaml` file. For
+  /// example a task that is executed every 5 minutes might have a max execution
+  /// time of 5 minutes.
+  @protected
+  Duration get maxExecutionTime => const Duration(minutes: 60);
+
+  /// Name of the cache key used for this handler.
+  @protected
+  String get cacheKey => '$runtimeType';
+
+  @override
+  @nonVirtual
+  Future<Body> get() async {
+    if (allowOnlyAppEngineCronAccess &&
+        request!.headers.value(xAppengineCron) != 'true') {
+      response!.statusCode = HttpStatus.unauthorized;
+      return Body.empty;
+    }
+
+    // Lookup if this task is already executing.
+    final isExecuting = await _cache.getOrCreateWithLocking(
+      subCacheName,
+      cacheKey,
+      createFn: null,
+    );
+    if (isExecuting != null) {
+      // Skip running this task.
+      final started = DateTime.fromMillisecondsSinceEpoch(
+        isExecuting.buffer.asUint64List().first,
+      );
+      log.info(
+        'Ignoring request to $runtimeType, already running since ${started.toIso8601String()}.',
+      );
+      response!.statusCode = HttpStatus.accepted;
+    } else {
+      // Mark the task as running.
+      try {
+        final now = Uint64List(1)..[0] = _now().millisecondsSinceEpoch;
+        await _cache.setWithLocking(
+          subCacheName,
+          cacheKey,
+          now.buffer.asUint8List(),
+          ttl: maxExecutionTime,
+        );
```

Comment:
So setWithLocking isn't transactional? If two calls to this request handler come in at the same time to two different services, they could both pass line 77 and come down to this call which will succeed for both?  If this were firestore, it would be (start transaction; read from key; maybe write key; complete transaction).

Reply:
That is a good point, maybe this doesn't actually WAI.

Redis does support transactions:
https://redis.io/docs/latest/develop/interact/transactions/

... but the Redis client (neat_cache) does not support it.

Moving this back to draft, and I'll file a bug for Redis transactional support.